src/python_vali/src/PySurface.cpp

```diff
@@ -162,6 +162,70 @@ void Init_PySurface(py::module& m) {
 
   py::class_<Surface, shared_ptr<Surface>>(
       m, "Surface", "Image stored in vRAM. Consists of 1+ SurfacePlane(s).")
+      .def(py::init([](Pixel_Format format, uint32_t width, uint32_t height, int gpuID) {
+            return std::shared_ptr<Surface>(
+                Surface::Make(format, width, height, CudaResMgr::Instance().GetCtx(gpuID)));
+        }),
+        py::arg("format"), py::arg("width"), py::arg("height"),
+        py::arg("gpu_id"), py::return_value_policy::take_ownership,
+        R"pbdoc(
+        Constructor.
+
+        :param format: target pixel format
+        :param width: width in pixels
+        :param height: height in pixels
+        :param gpu_id: GPU to use.
+    )pbdoc")
+      .def(py::init([](Pixel_Format format, uint32_t width, uint32_t height, size_t ctx) {
+            return std::shared_ptr<Surface>(
+                Surface::Make(format, width, height, (CUcontext)ctx));
+        }),
+        py::arg("format"), py::arg("width"), py::arg("height"),
+        py::arg("context"), py::return_value_policy::take_ownership,
+        R"pbdoc(
+        Constructor.
+
+        :param format: target pixel format
+        :param width: width in pixels
+        :param height: height in pixels
+        :param context: CUDA context to use
+    )pbdoc")
+      .def(py::init([](const PyDecoder& decoder, optional<Pixel_Format> format, int gpuID) {
```

Comment:
Please consider taking `py::object` instead and check for `Width`, `Height` and `Format` attributes.
Other constructor overload which takes `PyNvEncoder` will become redundant.

Also, IMO optional pixel format is ambiguous. 
If I pass decoder as argument, why would I choose different pixel format? Just to pass 1 argument (decoder) instead of 2 (width, height) ?

Reply:
a/ I have tried `py:object` but did not figure out how to check attribute and cast types. Could you help on that? Meanwhile, PyDecoder use `PixelFormat()` and PyNvEncoder use `GetPixelFormat()`.

b/ Agree with your point that optinal pixel format is not good. We could remove it. However, is there any reason developers have to allocate surface in advance?

My application code is quite verbose

```python
cctx = ColorspaceConversionContext(pyDec.ColorSpace, pyDec.ColorRange)

surf_nv12 = Surface(pyDec, PixelFormat.NV12, gpu_id)
surf_rgb = Surface(pyDec, PixelFormat.RGB, gpu_id)
surf_planar = Surface(pyDec, PixelFormat.RGB_PLANAR, gpu_id)

to_rgb = PySurfaceConverter(PixelFormat.NV12, PixelFormat.RGB, gpu_id)
to_planar = PySurfaceConverter(PixelFormat.RGB, PixelFormat.RGB_PLANAR, gpu_id)

pyDec.DecodeSingleSurface(surf_nv12)
to_rgb.Run(surf_nv12, surf_rgb, cctx)
to_planar.Run(surf_rgb, surf_planar, cctx)
```

I would prefer it to be
```python
surf_nv12 = pyDec.DecodeSingleSurface()
surf_planar = surf_nv12.Convert(PixelFormat.RGB_PLANAR)
```

Reply:
Check out [pybind11 docs](https://pybind11.readthedocs.io/en/stable/reference.html#_CPPv47hasattr6handlePKc)

```cpp
#include <pybind11/pybind11.h>
namespace py = pybind11;

void check_attribute_format(py::object obj) {
    return py::hasattr(obj, "Format);
}
```